### ISSUE TITLE
[BUG] partition pruning doesnt work with merge operation

### ISSUE THREAD

**Issue by abintrd** [delta-io/delta#1563]
#### Steps to reproduce

deltaTable.as("oldData")
        .merge(
            finalDF.where("OP != 'd'")
                    .as("newData"),
            condition = "oldData.ID = newData.ID and oldData.created_date in ('null','2022-12','2023-01')")
        .whenMatched
        .updateAll()
        .whenNotMatched
        .insertAll()
        .execute()

#### Observed results

In the physical plan I see no indication of partition filter being used on partition field created_date instead it ends up in the SortMergeJoin +- SortMergeJoin [ID#43253], [ID#43121], FullOuter, created_date#43136 IN (null,2022-12,2023-01)

### Environment information

* Delta Lake version:  2.1.0
* Spark version: 3.3.0
* Scala version: 2.12

### Willingness to contribute

The Delta Lake Community encourages bug fix contributions. Would you or another member of your organization be willing to contribute a fix for this bug to the Delta Lake code base?

- [ ] Yes. I can contribute a fix for this bug independently.
- [ *] Yes. I would be willing to contribute a fix for this bug with guidance from the Delta Lake community.
- [ ] No. I cannot contribute a bug fix at this time.

**Comment by scottsand-db** [comment]
Hi @abintrd - could you give us a succinct reproducible example? You've given some code about, but if you could also give some sample `data` that would let us easily reproduce this bug, that would help a lot!

**Comment by Ward1994** [comment]
hi, I encountered this but as well.
I think it can be easily reproduced when your condition has multiable column.
like `source.A = target.A and source.B = target.B and source.partitionKey =  target.partitionKey`

**Comment by Hannes237** [comment]
Hey!

I got the same issue taking a delta table `dt` which is partitioned using columns `a` (int) and `b` (date). The size of the table is ~12TB in a compressed state. I'm trying to merge a spark dataframe `df` which is distributed over multiple table partitions.

When using `.merge` no dynamic partition pruning is triggered. This leads to a high disk spill and makes the individual jobs take way longer than expected. The merge is more or less a LeftAnti Join of `df` on the existing `dt` using columns `a`, `b` and `c` (double). 

`
dt.alias('t').merge(
    source=df.alias('s'),
    condition=((F.col('t.a') == F.col('s.a')) & (F.col('t.b') == F.col('s.b')) & (F.col('t.c') == F.col('s.c')))
).whenNotMatchedInsertAll().execute()
`

I tried broadcasting `df` as suggested by several medium [posts (e.g. https://medium.com/@nethaji.bhuma/spark-dynamic-partition-pruning-a9777916293d)](https://medium.com/@nethaji.bhuma/spark-dynamic-partition-pruning-a9777916293d) and also enabling photon on Databricks.

`
dt.alias('t').merge(
    source=F.broadcast(df).alias('s'),
    condition=((F.col('t.a') == F.col('s.a')) & (F.col('t.b') == F.col('s.b')) & (F.col('t.c') == F.col('s.c')))
).whenNotMatchedInsertAll().execute()
`

Here an example plan.

> == Physical Plan ==
AdaptiveSparkPlan (62)
+- == Current Plan ==
   Execute WriteIntoDeltaCommand (32)
   +- WriteFiles (31)
      +- Sort (30)
         +- ShuffleQueryStage (29), Statistics(sizeInBytes=164.1 GiB, rowCount=7.13E+8, ColumnStat: N/A, isRuntime=true)
            +- Exchange (28)
               +- * Project (27)
                  +- * SortMergeJoin LeftAnti (26)
                     :- Sort (15)
                     :  +- ShuffleQueryStage (14), Statistics(sizeInBytes=302.9 GiB, rowCount=7.13E+8, ColumnStat: N/A, isRuntime=true)
                     :     +- Exchange (13)
                     :        +- * Filter (12)
                     :           +- SortAggregate (11)
                     :              +- Sort (10)
                     :                 +- ShuffleQueryStage (9), Statistics(sizeInBytes=579.2 GiB, rowCount=7.13E+8, ColumnStat: N/A, isRuntime=true)
                     :                    +- Exchange (8)
                     :                       +- SortAggregate (7)
                     :                          +- * Sort (6)
                     :                             +- * Project (5)
                     :                                +- * Project (4)
                     :                                   +- * Project (3)
                     :                                      +- * ColumnarToRow (2)
                     :                                         +- Scan parquet  (1)
                     +- * ColumnarToRow (25)
                        +- PhotonResultStage (24)
                           +- PhotonProject (23)
                              +- PhotonSort (22)
                                 +- PhotonProject (21)
                                    +- PhotonShuffleExchangeSource (20)
                                       +- ShuffleQueryStage (19), Statistics(sizeInBytes=7.3 TiB, rowCount=4.01E+11, ColumnStat: N/A, isRuntime=true)
                                          +- PhotonShuffleMapStage (18)
                                             +- PhotonShuffleExchangeSink (17)
                                                +- PhotonScan parquet unity_catalog.schema.table (16)

**Comment by Ward1994** [comment]
Hi, I think maybe you can consider use join to implement the merge which you can control partition pruning.

**Comment by Hannes237** [comment]
This would work for a simple appending insert, but would be not feasible for an upsert to a table.

For a simple append I got following code for inserting unmatched entries only:

``` python
deduplicated_batch = df.drop_duplicates(subset=join_columns)

remote_duplicates_df = target_table.toDF().join(
    F.broadcast(deduplicated_batch), on=join_columns, how="leftsemi"
)

insert_df = deduplicated_batch.join(
    F.broadcast(remote_duplicates_df), on=join_columns, how="leftanti"
)

insert_df.write.format("delta").mode("append").option("mergeSchema", "true").saveAsTable(table_name)
``` 

I would expect *delta* to handle this on its own!

**Comment by Ward1994** [comment]
Agree, maybe you can provider some sample data to help developer to debug this issue.

**Comment by foss-contributor** [comment]
@abintrd, thank you for your contribution! This issue has been inactive for over 18 months and will be closed in one week. To keep it open, please leave a comment on this issue or contact us in the `#contributors` channel on [Delta Lake Slack](https://go.delta.io/slack). We appreciate your involvement in Delta Lake!
